{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d03346a-f91a-4e97-8332-84b7ceb4dd72",
   "metadata": {},
   "outputs": [],
   "source": [
    "###  RNA-seq data analysis in R ###\n",
    "##  Основные шаги:\n",
    "# 1. Загрузка таблицы каунтов и conditions\n",
    "# 2. Аннотирование генов\n",
    "# 3. Фильтрация от NA и низкоэкспрессированных генов\n",
    "# 4.1 Нормализация (log2+quantile или vst/rlog) и визуализация (PCA plot, clustering, heatmap)\n",
    "# 4.2 Нормализация в DESeq2 для выявления DEGs\n",
    "# 5.1 limma\n",
    "# 5.2 DESeq2\n",
    "# 6. Анализ сигнальных путей на DEGs\n",
    "# 7. GSEA pathway analysis\n",
    "# 8. Отбор сигнальных путей по p.adj\n",
    "# 9. GSEA plots visualization\n",
    "# # \n",
    "\n",
    "# Установка библиотек:\n",
    "# 1) install.packages(\"pkg_name\")\n",
    "\n",
    "# 2) Если библиотека из Bioconductor:\n",
    "# if (!require(\"BiocManager\", quietly = TRUE))\n",
    "# install.packages(\"BiocManager\")\n",
    "# BiocManager::install(\"pkg_name\")\n",
    "#\n",
    "# 3) Можно установить напрямую, скачав файл пакета\n",
    "#\n",
    "#Активация нужных библиотек (должны быть установлены заранее):\n",
    "\n",
    "library(data.table)\n",
    "library(DESeq2)\n",
    "library(limma)\n",
    "library(fgsea)\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "library(\"AnnotationDbi\")\n",
    "library(devtools)\n",
    "library(dplyr)\n",
    "library(\"org.Mm.eg.db\")\n",
    "#BiocManager::install(\"phantasus\")\n",
    "library(phantasus)\n",
    "library(msigdbr)\n",
    "library(clusterProfiler)\n",
    "library(VennDiagram)\n",
    "library(pheatmap)\n",
    "library(utils)\n",
    "library(EnhancedVolcano)\n",
    "library(ggbeeswarm)\n",
    "\n",
    "library(devtools)\n",
    "library(rUtils)\n",
    "library(ggrepel)\n",
    "library(VennDiagram)\n",
    "#BiocManager::install(\"pathview\")\n",
    "library(pathview)\n",
    "\n",
    "\n",
    "# Загрузка таблицы экспрессии из базы данных GEO\n",
    "# library(GEOquery)\n",
    "# dir.create(\"cache\")\n",
    "# \n",
    "# ExpressionSet object:\n",
    "# - Expression matrix – таблица каунтов – exprs(es)\n",
    "# - fData – аннотация строк (название генов в разном формате)\n",
    "# - pData – аннотация столбцов (conditions)\n",
    "# \n",
    "# es <- getGEO(\"GSE53986\", AnnotGPL = TRUE, destdir=\"cache\")[[1]]\n",
    "# str(experimentData(es))\n",
    "# str(pData(es))       #таблица условий\n",
    "# head(fData(es))\n",
    "# fdata <-fData(es)\n",
    "# fdata                #посмотрим на струкутру не с помощью head\n",
    "# es$`treatment:ch1`   #образцы\n",
    "\n",
    "\n",
    "# В папке с проектом открываем таблицу каунтов и условий к ней\n",
    "\n",
    "#library(utils)\n",
    "table <- read.table('./data/GSE150365_LMNA_raw_counts.txt', sep = '\\t', header = T)\n",
    "cond <- read.csv(\"./data/conditions.tsv\", header=T, sep = '\\t', row.names = 1)\n",
    "\n",
    "\n",
    "# Удаляем букву \"X\" из названия образцов (колонки таблицы table). \n",
    "#Буква \"X\" осталась после подсчета каунтов.\n",
    "tags <- sapply(colnames(table), function (x) sub(\"X\", \"\", basename(x)))\n",
    "colnames(table) <- tags\n",
    "\n",
    "#Из всех условий выбраем миогенную дифференцировку (HS) и не дифференцированные клетки (control)\n",
    "cond <- cond[which(cond$Treatment == \"control\" | cond$Treatment==\"HS\"),]\n",
    "\n",
    "#Сделаем одинаковый порядок образцов в table (это название столбцов) и в cond \n",
    "#(названия строк).\n",
    "#Это важно для запуска анализа дифф. экспрессии\n",
    "table[,-1] <- table[,rownames(cond)]\n",
    "\n",
    "\n",
    "# Посмотрим статистику\n",
    "\n",
    "col_sums <- colSums(table[,-1])\n",
    "col_sums #суммарное количество ридов на образец\n",
    "plot(col_sums, xlab=\"sample\") #простой график количества ридов\n",
    "mean(col_sums) #среднее количество ридов во всем эксперименте\n",
    "col_sums[col_sums < 7500000] #образцы, у которых меньше 7.5 миллиона ридов\n",
    "boxplot(table[,c(2:9)]) #boxplot для выбранных образцов. Много генов с нулевым количеством каунтов.\n",
    "#нужно делать фильтрацию\n",
    "\n",
    "# Аннотирование строк\n",
    "#table$Gene_id <- sub(\"\\\\..*\", \"\", rownames(table)) #удаление .1 в ENSEMBL ID, если есть\n",
    "\n",
    "table <- unique(table, by = \"Gene_id_ENSEMBL\") #оставляем гены с уникальным ensembl id \n",
    "#(если есть повторы, но не должно быть)\n",
    "#library(data.table)\n",
    "table <- as.data.table(table)\n",
    "#Переводим ensembl в symbol\n",
    "#library(\"AnnotationDbi\")\n",
    "#library(\"org.Mm.eg.db\") #аннотация генов для мыши\n",
    "table[, Symbol:= mapIds(org.Mm.eg.db, keys=Gene_id_ENSEMBL, \n",
    "                         keytype=\"ENSEMBL\", column=\"SYMBOL\")] #новый столбец symbol \n",
    "                                                              # по ENSEMBL id\n",
    "#В столбце symbol есть NA значения. Нужно будет их потом убрать.\n",
    "\n",
    "#Возможные типы клучей в базе данных\n",
    "keytypes(org.Mm.eg.db)\n",
    "\n",
    "\n",
    "# Можно сделать Expression Set object для экспорта в Phantasus\n",
    "es <- ExpressionSet(as.matrix(table[,!c(\"Gene_id_ENSEMBL\", \"Symbol\")])) #только каунты \n",
    "head(exprs(es),20)\n",
    "fData(es) <- table[,c(\"Gene_id_ENSEMBL\", \"Symbol\")] #название гена в различных базах данных\n",
    "head(fData(es))\n",
    "rownames(es) <- fData(es)$Gene_id_ENSEMBL #в качестве строк сделаем ensembl id\n",
    "pData(es) <- cond\n",
    "head(pData(es)) #таблица условий\n",
    "# BiocManager::install(\"phantasus\")\n",
    "# library(phantasus)\n",
    "\n",
    "write.gct(es,\"es_raw.gct\")\n",
    "\n",
    "\n",
    "# Посмотрим на экспрессию гена бета-актина\n",
    "exprs(es)[which(fData(es)$Symbol == \"Actb\"), ]\n",
    "#Быстро сохраним график в формате png. Если нужно поменять оформление графика, \n",
    "#то лучше использовать пакет ggplot2\n",
    "png(\"Actb.png\", 1000, 800)\n",
    "plot(exprs(es)[which(fData(es)$Symbol == \"Actb\"), ])\n",
    "dev.off()\n",
    "\n",
    "\n",
    "# Фильтрация. Удаляем NA и дибликаты в symbol\n",
    "table <- table[!is.na(table$Symbol)]   # удаляем строки с NA в symbol\n",
    "length(unique(table$Symbol))   # Проверяем есть ли дубликаты. \n",
    "#Есть, так как всего строк в table больше 32000, а уникальных из них 31950\n",
    "\n",
    "# table <- unique(table, by=\"Symbol\") # Можно удалить дубликаты так, \n",
    "                                       #но unique оставляет рандомно строки\n",
    "\n",
    "\n",
    "# Дедубликация. Разным ensembl может соответствовать один symbol\n",
    "\n",
    "# Для каждого ensembl id заменим экспрессию гена на медианное значение\n",
    "#library(dplyr) #чтобы заработала функция %>% \n",
    "table_dedup <- table[,!c(\"Gene_id_ENSEMBL\")] %>% group_by(Symbol) %>% \n",
    "                summarise_all(median) %>% as.data.table()  #выполняется долго\n",
    "\n",
    "table_dedup[, Gene_id := table[match(table_dedup$Symbol, table$Symbol), \"Gene_id_ENSEMBL\"] ]  \n",
    "# Добавили Gene_id ENSEMBL, она была удалена\n",
    "table_dedup[, Entrez := mapIds(org.Mm.eg.db, keys=Symbol, keytype=\"SYMBOL\", column=\"ENTREZID\")] \n",
    "# Добавили Entrez id\n",
    "table_dedup$Entrez <- as.character(table_dedup$Entrez) %>% replace(list = table_dedup$Entrez == \"NULL\", values = NA)  \n",
    "# Заменили NULL на NA,если нет entrez \n",
    "table_dedup <- table_dedup[!is.na(table_dedup$Entrez)] # Remove Entrez NA\n",
    "\n",
    "\n",
    "# ЦЕЛЬ: необходимо статистически сравнить средние экспрессий между двумя\n",
    "# выборками образцов. Нулевая гипотеза говорит, что уровень экспрессии не будет различаться.\n",
    "# \n",
    "# Возможные решения в классической статистике:\n",
    "#   - т-тест (не подходит, так как не нормальное распределение данных)\n",
    "#   - Манн-Уитни (слишком слабый, так как чаще всего мало образцов в группе)\n",
    "# \n",
    "# Они нам не подходят.\n",
    "# Для использования стат. критериев нужно оценить экспрессию и дисперсию для каждого гена.\n",
    "# Для этого используют различные модели. \n",
    "# \n",
    "# Далее три пути анализа:\n",
    "#   \n",
    "# 1. Limma. На вход подаются нормализованные данные. \n",
    "# Использует линейные модели для моделирования экспрессии генов и их дисперсии. \n",
    "# Предполагается нормальность распределения данных для оценки коэффициентов модели. \n",
    "# Исходя из \"одинаковости\" коэффициентов модели перед факторами делают вывод об изменении экспрессии гена.\n",
    "# \n",
    "# 2. DESeq2. На вход подаются \"сырые\" каунты, но после фильтрации. \n",
    "# Использует Generalized linear model (GLM) для моделирования экспрессии генов и их дисперсии.\n",
    "# При оцнке коэффициентов модели используют отрицательное биномиальное распределение,\n",
    "# которое более точно оченивает экспрессию. \n",
    "# \n",
    "# 3. EdgeR (не будем разбирать в рамках этого курса). \n",
    "# На вход подаются нормализованные данные. Тоже используется отрицательное биномиальное распределение.\n",
    "# \n",
    "# \n",
    "# При анализе в модель включают потенциальные факторы, такие как бач-эффект или \n",
    "# экспериментальные условия, которые могут влиять на экспрессию генов. \n",
    "# Поэтому получается более точная оценка истинных дифференциальных паттернов экспрессии.\n",
    "# \n",
    "# Достоверность определения коэффициентов линейной модели (p-value) в DESeq2 оценивается\n",
    "# с помощью теста Вальда. \n",
    "# Если p-значение мало, мы отвергаем нулевую гипотезу и утверждаем, что ген дифференциально экспрессирован.\n",
    "\n",
    "\n",
    "#Сделаем анализ дифференциальной экспрессии с помощью Limma\n",
    "\n",
    "# # Профильтруем гены. Оставим только те, у которых среднее значение экспресии больше 1\n",
    "table_dedup$mean <- rowMeans(table_dedup[,!c(\"Gene_id\", \"Symbol\", \"Entrez\")])\n",
    "table_dedup <- table_dedup[mean > 1,]\n",
    "table_dedup$mean <- NULL # Удалим колонку \"mean\"\n",
    "\n",
    "### Получили 15376 генов\n",
    "# Сохраним\n",
    "es <- ExpressionSet(as.matrix(table_dedup[,!c(\"Gene_id\", \"Symbol\", \"Entrez\")]))\n",
    "fData(es) <- table_dedup[, c(\"Gene_id\", \"Symbol\", \"Entrez\")]\n",
    "rownames(es) <- fData(es)$Gene_id\n",
    "pData(es) <- cond\n",
    "write.gct(es,\"es_filter.gct\")\n",
    "\n",
    "\n",
    "# Принято использовать нормализацию log2+1, а потом quantile нормализацию\n",
    "# Как было в phantasus\n",
    "es.qnorm.top12K <- es       #скопировали объект es в новую переменную\n",
    "#library(limma)\n",
    "exprs(es.qnorm.top12K) <- normalizeBetweenArrays(log2(exprs(es.qnorm.top12K) + 1), method=\"quantile\")  \n",
    "# log2 and quantile normalization\n",
    "#логарифмические отношения в среднем равнялись одному значению в каждом образце \n",
    "\n",
    "es.qnorm.top12K <- es.qnorm.top12K[head(order(apply(exprs(es.qnorm.top12K), 1, mean), \n",
    "                                              decreasing = TRUE), 12000), ]\n",
    "#топ 12000 генов по среднему значению экспрессии\n",
    "\n",
    "fData(es.qnorm.top12K)$mean <- apply(exprs(es.qnorm.top12K), 1, mean)  \n",
    "# записываем среднее в аннотацию строк\n",
    "\n",
    "write.gct(es.qnorm.top12K, file=\"./es.qnorm.top12k.gct\")  #сохраним\n",
    "\n",
    "\n",
    "## Анализ главных компонент (PCA plot)\n",
    "\n",
    "# 1 способ\n",
    "pca <- prcomp(t(exprs(es.qnorm.top12K)))\n",
    "plot(pca$x[, 1:2])\n",
    "\n",
    "# 2 способ\n",
    "#install.packages(\"devtools\")\n",
    "#devtools::install_github(\"assaron/r-utils\", force = T)\n",
    "#library(rUtils)\n",
    "#library(ggrepel)\n",
    "\n",
    "p <- pcaPlot(es.qnorm.top12K, 1, 2) + \n",
    "  geom_text_repel(aes(label=samples)) +\n",
    "  aes(color=Condition)\n",
    "print(p)         # PCA plot colored for conditions\n",
    "#другой способ сохранить график, если он сделан с помощью ggplot2\n",
    "ggsave(p, width=12, height=8, file=\"PCA_1.png\")\n",
    "\n",
    "p <- pcaPlot(es.qnorm.top12K, 1, 2) + \n",
    "  geom_text_repel(aes(label=samples)) +\n",
    "  aes(color=Day)+\n",
    "  stat_ellipse(aes(color=Day))\n",
    "print(p)           # Colored by Day of differentiation\n",
    "ggsave(p, width=12, height=8, file=\"PCA_2.png\")\n",
    "\n",
    "\n",
    "\n",
    "# Кластеризация: hierarchical and k-means\n",
    "#library(pheatmap)\n",
    "# z-score нормализация. Среднее значение всех значений равно 0, а стандартное отклонение равно 1.\n",
    "es.qnorm.top12K.z <- es.qnorm.top12K\n",
    "exprs(es.qnorm.top12K.z) <- t(apply(exprs(es.qnorm.top12K.z),1,scale)) # транспонировали\n",
    "phmap_z <- pheatmap(exprs(es.qnorm.top12K.z), \n",
    "                    kmeans_k=5, #количество кластеров определяется методом \"логтя\" или \"колена\"\n",
    "                    clustering_distance_cols = \"correlation\", \n",
    "                    clustering_method = \"average\",\n",
    "                    annotation_col = cond,\n",
    "                    color=colorRampPalette(c(\"navy\", \"white\", \"red\"))(50))\n",
    "head(phmap_z$kmeans$cluster, 20)  # какие гены к каким кластерам принадлежат\n",
    "\n",
    "#Здесь выбрали количество кластеров сами, но более правильно использовать метод \"логтя\" или \"колена\"\n",
    "\n",
    "\n",
    "## Удалим выбросы среди образцов (выявили с помощью phantasus):\n",
    "outliers <- c(\"wt_HS_term_1\",\"482_HS_2d_1\")\n",
    "es.qnorm.top12K.no_out <- es.qnorm.top12K[, !(colnames(es.qnorm.top12K) %in% outliers)]\n",
    "colnames(es.qnorm.top12K.no_out)\n",
    "\n",
    "\n",
    "\n",
    "## Удалим их из сырых данных и обновим ExpressionSet\n",
    "# Удалим образцы из \"table_dedup\" (unnormalized table!)\n",
    "table_dedup <- as.data.frame(table_dedup)\n",
    "table_dedup <- table_dedup[,!(names(table_dedup) %in% outliers)] %>% as.data.table()\n",
    "cond <- cond[!(rownames(cond) %in% outliers),]\n",
    "\n",
    "#Новый Expression Set\n",
    "es <- ExpressionSet(as.matrix(table_dedup[, !c(\"Gene_id\", \"Symbol\", \"Entrez\")]))\n",
    "fData(es) <- table_dedup[, c(\"Gene_id\", \"Symbol\", \"Entrez\")]\n",
    "rownames(es) <- fData(es)$Gene_id\n",
    "pData(es) <- cond\n",
    "write.gct(es, \"es_filt_no_out.gct\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Differential Expression Analysis (DEA) использую Limma\n",
    "#Анализ дифференциальной экспрессии\n",
    "#выбираем, какой столбец будет спользован для разбиения на группы сравнения\n",
    "#В основе - линейная модель (линейная регрессия), где мы хотим предсказать уровень экспрессии каждого гена. \n",
    "#В зависимости от условий.\n",
    "#\n",
    "es.design <- model.matrix(~0+Condition, data=pData(es.qnorm.top12K.no_out)) \n",
    "es.design\n",
    "#library(limma)\n",
    "fit <- lmFit(es.qnorm.top12K.no_out, es.design)\n",
    "\n",
    "fit1 <- contrasts.fit(fit, makeContrasts2(c(\"Condition\", \"232_control\", \"wt_control\"),\n",
    "                                          levels=es.design)) #парное сравнение\n",
    "fit1 <- eBayes(fit1)\n",
    "de  <- topTable(fit1, adjust.method=\"BH\", number=Inf) #поправка на множественные сравнения\n",
    "head(de)\n",
    "de <- as.data.table(de, keep.rownames=TRUE)\n",
    "de  <- de[order(t, decreasing = T)]\n",
    "de$logFC <- as.numeric(de$logFC)\n",
    "de_filt <- de[adj.P.Val<0.05 & (logFC > 1.5 | logFC < -1.5)] \n",
    "# Отобрали дифф. экспрессирующиеся гены (DEGs)\n",
    "\n",
    "#Построим Volcanoplot\n",
    "volcanoplot(fit1, style = \"p-value\", highlight = 116, names = fit1$genes$Symbol,\n",
    "            xlab = \"Log2 Fold Change\", ylab = NULL, pch=16, cex=0.35)\n",
    "\n",
    "# Более красивый график\n",
    "#library(EnhancedVolcano)\n",
    "volcano_limma <- EnhancedVolcano(de, title = \"Volcanoplot_Limma\",\n",
    "                lab = fit1$genes$Symbol, #ylim = c(0,10),\n",
    "                x = 'logFC',\n",
    "                y = 'adj.P.Val', pCutoff = 0.05,\n",
    "                FCcutoff = 1.5,)\n",
    "ggsave(volcano_limma, width=10, height=10, file=\"Volcanoplot_Limma.png\")\n",
    "\n",
    "# Другое парное сравнение \"482_control\" vs \"wt_control\" \n",
    "\n",
    "fit2      <- contrasts.fit(fit, makeContrasts2(c(\"Condition\", \"482_control\", \"wt_control\"),\n",
    "                                               levels=es.design))\n",
    "fit2      <- eBayes(fit2)\n",
    "de2        <- topTable(fit2, adjust.method=\"BH\", number=Inf)\n",
    "head(de2)\n",
    "de2        <- as.data.table(de2, keep.rownames=TRUE)\n",
    "de2        <- de2[order(t, decreasing = T)]\n",
    "de2_filt   <- de2[adj.P.Val<0.05 & (logFC > 1.5 | logFC < -1.5)] #DEGs\n",
    "\n",
    "# Пересечение upregulated DEGs двух сравненний с помощью диаграмм Венна\n",
    "#library(VennDiagram)\n",
    "set1 <- de_filt[logFC > 1.5, Symbol] \n",
    "set2 <- de2_filt[logFC > 1.5, Symbol] \n",
    "venn.diagram(\n",
    "  x = list(set1, set2),\n",
    "  category.names = c(\"232_vs_wt\" , \"482_vs_wt\"),\n",
    "  filename = 'venn_diagramm.png',\n",
    "  output=TRUE, \n",
    "  height = 1000, \n",
    "  width = 1000, \n",
    "  resolution =300, \n",
    "  fill = c(\"#E69F00\", \"#56B4E9\"),\n",
    "  lty = 'blank',\n",
    "  cat.cex = 0.6,\n",
    "  cat.fontface = \"bold\",\n",
    "  cat.default.pos = \"outer\",\n",
    "  cat.pos = c(0, -27),\n",
    "  cat.dist = c(0.01, 0.02),\n",
    "  cat.fontfamily = \"sans\")\n",
    "#сохранили в файл venn_diagramm.png\n",
    "\n",
    "intersect(set1, set2) # Пересечение upregulated DEGs\n",
    "setdiff(set2, set1) # Уникальные upregulated DEGs для 482_vs_WT\n",
    "\n",
    "\n",
    "# pathway analysis with fgsea\n",
    "stats <- de[, setNames(t, Entrez)] \n",
    "#выбрали из таблицы de только столбцы t и названия генов в столбце Entrez\n",
    "\n",
    "#library(msigdbr)\n",
    "# GO BP pathways from MSigDB\n",
    "m_df <- msigdbr(species = \"Mus musculus\", category = \"C5\", subcategory = \"BP\")\n",
    "m_df\n",
    "pathways <- split(m_df$entrez_gene, m_df$gs_name) #entrez id и название пути\n",
    "#подгрузили базу к себе \n",
    "\n",
    "#library(fgsea)\n",
    "#пересекаем наши гены с генами из базы\n",
    "frML <- fgseaMultilevel(pathways, stats,\n",
    "                          sampleSize = 100,\n",
    "                          nproc=4, minSize=15, maxSize=500)\n",
    "frML <- frML[order(padj)] \n",
    "frML <- frML[padj < 0.05]\n",
    "\n",
    "collapsed_pathways <- collapsePathways(frML, pathways, stats)\n",
    "#схлопываем похожие сигнальные пути\n",
    "str(collapsed_pathways) \n",
    "\n",
    "main_pathways <- frML[pathway %in% collapsed_pathways$mainPathways][\n",
    "  order(sign(ES)*log(pval)), pathway] # отбираем только main pathways по значению ES и p-value\n",
    "\n",
    "frMain <- frML[match(main_pathways, pathway)] #оставляем только main пути\n",
    "frMain[, leadingEdge := lapply(leadingEdge, mapIds, \n",
    "                                 x=org.Mm.eg.db, keytype=\"ENTREZID\", column=\"SYMBOL\")] \n",
    "#переименуем гены из Entrez id в Symbol\n",
    "#добавили столбец symbol, содержащий название генов, участвующих в данном сигнальном пути\n",
    "\n",
    "dir.create(\"gsea\") #сделали папку\n",
    "fwrite(frMain, file=\"gsea/FGSEA_filtered.tsv\", sep=\"\\t\", sep2=c(\"\", \" \", \"\")) \n",
    "# сохранили в виде tsv файла. Можно в другом формате.\n",
    "\n",
    "\n",
    "# Построим GSEA plot для определенного сигнального пути\n",
    "gsea_plot <- plotEnrichment(pathways[[\"GOBP_DNA_REPLICATION\"]], stats) + \n",
    "  ggtitle(\"GOBP_DNA_REPLICATION\")\n",
    "ggsave(gsea_plot, width=8, height=4, file=\"gsea/GOBP_DNA_REPLICATION.png\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Поиск дифференциально экспрессированных генов с помощью DESeq2\n",
    "## На вход подается фильтрованная и не нормированная таблица каунтов\n",
    "## Возьмем для анализа 15000 генов (можно все)\n",
    "\n",
    "\n",
    "library(DESeq2)\n",
    "exprs(es) <- apply(exprs(es), 2, as.integer) #Взяли уже имеющийся датасет es, сделали тип integer\n",
    "\n",
    "# Внутри DESeq2 своя нормализация, основанная на медианной экспрессии. Сейчас ее сделаем.\n",
    "dds <- DESeqDataSetFromMatrix(countData = exprs(es), \n",
    "                              colData = pData(es),\n",
    "                              design=~Condition)     # Condition это наши группы сравнения\n",
    "dds <- dds[rowSums(counts(dds)) > 10, ] #отфильтруем гены, у которых в сумме по всем образцам меньше 10 каунтов\n",
    "dds$Condition <- relevel(dds$Condition, ref = \"wt_control\") #указали группу с \"базовым\" уровнем экспрессии\n",
    "dds <- DESeq(dds) #сам DESeq\n",
    "dds\n",
    "\n",
    "\n",
    "# Построим PCA plot\n",
    "#Для начала нужно сделать vst нормализацию данных\n",
    "vst <- varianceStabilizingTransformation(dds) # эта функция работает на основе dds\n",
    "\n",
    "plotPCA(vst,intgroup=c(\"Condition\")) + \n",
    "  geom_text_repel(aes(label=name))+\n",
    "  coord_fixed(ratio = 4)\n",
    "\n",
    "plotPCA(vst,intgroup=c(\"Day\")) + coord_fixed(ratio = 4)+\n",
    "  geom_text_repel(aes(label=name))+\n",
    "  stat_ellipse(aes(color=Day))\n",
    "\n",
    "plotPCA(vst,intgroup=c(\"Cell_type\")) + \n",
    "  coord_fixed(ratio = 4)+\n",
    "  geom_text_repel(aes(label=name))\n",
    "\n",
    "ggsave(\"PCA_plot_DESeq2.png\") #если хотите сохранить график\n",
    "\n",
    "#Иерархическая кластеризация образцов\n",
    "dists <-  dist(t(assay(vst)))\n",
    "plot(hclust(dists)) # иерархическая кластеризация образцов\n",
    "#трипликаты вместе\n",
    "\n",
    "# Построим тепловую карту 20 самых вариабельных генов\n",
    "deseq_counts <- assay(vst) \n",
    "\n",
    "deseq_counts.z <- t(apply(deseq_counts,1,scale)) # scaling\n",
    "colnames(deseq_counts.z) <- colnames(deseq_counts)\n",
    "deseq_counts.z <- deseq_counts.z[c(1:20),] #топ-20 генов (для удобства и быстроты)\n",
    "#можно здесь выбрать интересующие вас гены\n",
    "pheatmap(deseq_counts.z, \n",
    "         #kmeans_k=15,                      \n",
    "         clustering_distance_cols = \"correlation\", \n",
    "         clustering_method = \"average\",\n",
    "         annotation_col = cond,\n",
    "         color=colorRampPalette(c(\"navy\", \"white\", \"red\"))(50)) #50 шаг цветовой градации\n",
    "#можно сделать k-means кластеризацию\n",
    "\n",
    "\n",
    "# Найдем дифференциально экспрессирующиеся гены 232_control vs wt_control\n",
    "dir.create(\"./deseq2/\", showWarnings = F)  #output directory\n",
    "unique(dds$Condition)\n",
    "# Важен порядок написания образцов в contrasts\n",
    "#На первом месте указываем \"воздействие\", а на стором то, с чем сравниваем, т.е. контроль\n",
    "# log2FC = 232_control -  wt_control   \n",
    "# If  log2FC>0 => gene is upregulated in 232_control; if log2FC<0 => gene is downregulated\n",
    "\n",
    "# Уточним группы сравнения\n",
    "deseq2_res <- results(dds, contrast = c(\"Condition\", \"232_control\", \"wt_control\"), cooksCutoff = F)\n",
    "head(deseq2_res)\n",
    "deseq2_res <- data.table(ID=rownames(deseq2_res), as.data.table(deseq2_res))  \n",
    "head(deseq2_res)\n",
    "tail(deseq2_res)\n",
    "\n",
    "\n",
    "head(fData(es)) #разные аннотации генов (ensembl, symbol, entrez)\n",
    "deseq2_res <- cbind(deseq2_res, fData(es)) # добавили аннотацию таблице дифф. генов\n",
    "deseq2_res <- deseq2_res[order(stat, decreasing = T), ] #сортировка по переменной stat\n",
    "#stat - статистика Вальда\n",
    "head(deseq2_res)\n",
    "\n",
    "# Myh7 ген\n",
    "deseq2_res[Symbol == \"Myh7\"]\n",
    "\n",
    "#library(ggbeeswarm)\n",
    "geneCounts <- plotCounts(dds, gene = \"ENSMUSG00000053093\", intgroup=c(\"Condition\"),returnData = TRUE)\n",
    "ggplot(geneCounts, aes(x = Condition, y = count, colour = Condition, size = 3)) +\n",
    "  scale_y_log10() +  geom_beeswarm(cex = 3) #+ geom_point(size=3)\n",
    "\n",
    "# Сохраним табличку\n",
    "fwrite(deseq2_res, file=\"./deseq2/232_control_vs_wt_control.de.tsv\", sep=\"\\t\")\n",
    "\n",
    "# Volcanoplot\n",
    "\n",
    "#library(EnhancedVolcano)\n",
    "#png(\"Volcanoplot_232_contr_vs_wt_contr.png\", 800, 800, pointsize=30)\n",
    "Volcano <- EnhancedVolcano(deseq2_res, title = \"Volcanoplot_DESeq2\",\n",
    "                           lab = deseq2_res$Symbol, #ylim = c(0,10),\n",
    "                           x = 'log2FoldChange',\n",
    "                           y = 'padj', pCutoff = 0.05,\n",
    "                           FCcutoff = 1.5,)\n",
    "Volcano\n",
    "#dev.off()\n",
    "\n",
    "\n",
    "## Pathway analysis\n",
    "\n",
    "## Сделаем анализ сигнальных путей на DEGs (Differentially Expressed Genes)\n",
    "## Будем использовать базу данных GO BP (Gene Ontology - Biological Processes)\n",
    "# UP and DOWN DEGs анализируются отдельно\n",
    "#library(clusterProfiler)\n",
    "\n",
    "deg_filt <- deseq2_res[padj<0.05 & (log2FoldChange > 1.5 | log2FoldChange < -1.5)]\n",
    "\n",
    "genes <- deg_filt[log2FoldChange > 1.5]$Symbol\n",
    "\n",
    "go_deg <- enrichGO(genes, 'org.Mm.eg.db', ont=\"BP\", pvalueCutoff=0.05, keyType = \"SYMBOL\")\n",
    "head(go_deg)\n",
    "head(go_deg$Description)\n",
    "\n",
    "barplot(go_deg, showCategory=10) # Первые 10 путей с p.adj < 0.05 \n",
    "dotplot(go_deg, showCategory=10)\n",
    "\n",
    "go_deg_res <- go_deg@result\n",
    "go_deg_res <- go_deg_res %>% dplyr::filter(go_deg_res$p.adjust < 0.05)\n",
    "fwrite(go_deg_res,\"EnrichGO_Up.csv\")\n",
    "\n",
    "\n",
    "# Можно использовать другие базы данных сигнальных путй (как в R, так и онлайн).\n",
    "# Из онлайн тулов можно использовать Msigdb и Enrichr\n",
    "\n",
    "# Теперь попоробуем KEGG базу данных для анализа сигнальных путей\n",
    "\n",
    "genes <- deg_filt[log2FoldChange > 1.5]$Entrez  #FC>1.5, p.adj=0.05\n",
    "kegg_deg <- enrichKEGG(genes, organism = \"mmu\", keyType = 'ncbi-geneid', pvalueCutoff = 0.05)\n",
    "barplot(kegg_deg, showCategory=10) # Первые 10 путей с p.adj < 0.05  \n",
    "\n",
    "#Визуализируем сигнальный путь Calcium signaling pathway\n",
    "#BiocManager::install(\"pathview\")\n",
    "#library(pathview)\n",
    "de_genes <- deseq2_res[abs(log2FoldChange) > 1.5 & padj < 0.05,]\n",
    "genes <- de_genes$log2FoldChange\n",
    "names(genes) <- de_genes$Entrez\n",
    "dme <- pathview(gene.data=genes, pathway.id=\"mmu04020\", species = \"mmu\")\n",
    "# Открываем сохраненные .png файлы\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## GSEA - Gene Set Enrichment Analysis\n",
    "\n",
    "#library(clusterProfiler)\n",
    "genes <- as.vector(deseq2_res$stat) #без фильтрации по p.adj и LFC\n",
    "names(genes) <- as.vector(deseq2_res$Symbol)\n",
    "genes <- sort(genes, decreasing = TRUE) #именованный список генов, отсортированных по stat\n",
    "gse <- gseGO(geneList     = genes,\n",
    "             OrgDb        = org.Mm.eg.db,\n",
    "             ont          = \"BP\",\n",
    "             keyType = \"SYMBOL\",\n",
    "             nPerm=10000)\n",
    "\n",
    "# Dot plot of GSEA\n",
    "dotplot(gse, showCategory=10, split=\".sign\") + facet_grid(.~.sign) #+ ggtitle(\"dotplot for GSEA\")\n",
    "gseaplot(gse, geneSetID = 5, by = \"runningScore\", title = gse$Description[5]) #визуализируем 5-ый сигнальный путь\n",
    "\n",
    "\n",
    "# Теперь используем fgsea package\n",
    "# Скачаем пути из базы данных\n",
    "\n",
    "#library(msigdbr)\n",
    "# Gene Ontology Biological Processes pathways from MSigDB (https://www.gsea-msigdb.org/gsea/msigdb/index.jsp)\n",
    "m_df <- msigdbr(species = \"Mus musculus\", category = \"C5\", subcategory = \"BP\")\n",
    "m_df\n",
    "pathways <- split(m_df$entrez_gene, m_df$gs_name)\n",
    "\n",
    "# t statistics\n",
    "stats <- deseq2_res[, setNames(stat, Entrez)]\n",
    "\n",
    "#library(fgsea)\n",
    "fr <- fgsea(pathways, stats, nperm = 100000, nproc=4, minSize=15, maxSize=500)\n",
    "fr[order(padj)]\n",
    "fr_res <- fr[order(NES)][padj < 0.05]\n",
    "\n",
    "#сделаем новый столбец, в котором будут указаны символы генов, участвующих в сигнальном пути\n",
    "for (val in 1:nrow(fr_res)){\n",
    "  fr_res$symbol[[val]] <- fr_res$leadingEdge[[val]] %>% mapIds(org.Mm.eg.db, keys = ., keytype = \"ENTREZID\",column = c('SYMBOL'), multiVals=\"first\") %>%  paste(collapse = \", \")\n",
    "  val = val+1\n",
    "} #выполняется долго\n",
    "\n",
    "# Сохраним\n",
    "dir.create(\"gsea\")\n",
    "fwrite(fr_res, file=\"gsea/fgsea_232_contr_vs_wt_contr.csv\", sep=\",\")\n",
    "\n",
    "# Построим GSEA plots\n",
    "topPathwaysUp <- fr_res[ES > 0 & padj < 0.05][head(order(padj), n=10), pathway]\n",
    "topPathwaysDown <- fr_res[ES < 0 & padj < 0.05][head(order(padj), n=10), pathway]\n",
    "topPathways <- c(topPathwaysUp, rev(topPathwaysDown))\n",
    "\n",
    "FGSEA <- plotGseaTable(pathways[topPathways], stats, fr_res, \n",
    "                       gseaParam=0.5)\n",
    "\n",
    "FGSEA\n",
    "ggsave(file=\"Gsea_plot_232_c_vs_wt_c.jpeg\", dpi=300, width = 14, height = 4)\n",
    "\n",
    "\n",
    "# GSEA plots \n",
    "\n",
    "plotEnrichment(pathways[[\"GOBP_MUSCLE_CONTRACTION\"]],\n",
    "               stats, ticksSize=0.5) + \n",
    "  labs(title=\"Muscle contraction\")+\n",
    "  theme(plot.title = element_text(size=30, color=\"black\"),\n",
    "        axis.text = element_text(color = \"black\", size = 18), \n",
    "        axis.title.y = element_text(size=25, color=\"black\"),\n",
    "        axis.title.x = element_text(size=25, color=\"black\")\n",
    "  )\n",
    "ggsave(file=\"Muscle_contraction.jpeg\", dpi=300, width = 8, height = 4)\n",
    "\n",
    "\n",
    "plotEnrichment(pathways[[\"GOBP_MITOCHONDRIAL_RESPIRATORY_CHAIN_COMPLEX_ASSEMBLY\"]],\n",
    "               stats, ticksSize=0.5) + \n",
    "  labs(title=\"MITOCHONDRIAL RESPIRATORY CHAIN COMPLEX ASSEMBLY\")+\n",
    "  theme(plot.title = element_text(size=15, color=\"black\"),\n",
    "        axis.text = element_text(color = \"black\", size = 18), \n",
    "        axis.title.y = element_text(size=25, color=\"black\"),\n",
    "        axis.title.x = element_text(size=25, color=\"black\")\n",
    "  )\n",
    "ggsave(file=\"Mitochondrial_respiration.jpeg\", dpi=300, width = 8, height = 4)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "417de2f0-c99b-4a57-8f24-4afe9bd1031b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.16 (BiocManager 1.30.22), R 4.2.3 (2023-03-15)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'DESeq2' 'AnnotationDbi'”\n",
      "Installing package(s) 'limma', 'fgsea', 'org.Mm.eg.db', 'phantasus',\n",
      "  'clusterProfiler', 'pathview'\n",
      "\n",
      "also installing the dependencies ‘prettyunits’, ‘progress’, ‘gridGraphics’, ‘tweenr’, ‘polyclip’, ‘RcppEigen’, ‘gridExtra’, ‘lazyeval’, ‘clipr’, ‘hms’, ‘vroom’, ‘tzdb’, ‘R.oo’, ‘R.methodsS3’, ‘highr’, ‘xfun’, ‘yaml’, ‘mvtnorm’, ‘DEoptimR’, ‘ggfun’, ‘ggplotify’, ‘patchwork’, ‘ggforce’, ‘viridis’, ‘tidygraph’, ‘graphlayouts’, ‘cpp11’, ‘ape’, ‘tidytree’, ‘treeio’, ‘readr’, ‘xml2’, ‘R.utils’, ‘brew’, ‘later’, ‘promises’, ‘knitr’, ‘remotes’, ‘webutils’, ‘rappdirs’, ‘zip’, ‘Rhdf5lib’, ‘rhdf5filters’, ‘stringi’, ‘systemfonts’, ‘pcaPP’, ‘robustbase’, ‘HDO.db’, ‘reshape2’, ‘aplot’, ‘ggnewscale’, ‘ggraph’, ‘igraph’, ‘purrr’, ‘scatterpie’, ‘shadowtext’, ‘ggtree’, ‘fs’, ‘data.table’, ‘cowplot’, ‘fastmatch’, ‘protolite’, ‘GEOquery’, ‘Rook’, ‘httpuv’, ‘opencpu’, ‘assertthat’, ‘rhdf5’, ‘stringr’, ‘svglite’, ‘ccaPP’, ‘downloader’, ‘DOSE’, ‘enrichplot’, ‘GO.db’, ‘GOSemSim’, ‘gson’, ‘plyr’, ‘qvalue’, ‘tidyr’, ‘yulab.utils’, ‘KEGGgraph’, ‘Rgraphviz’, ‘graph’, ‘org.Hs.eg.db’\n",
      "\n",
      "\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘polyclip’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘RcppEigen’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘lazyeval’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘xfun’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘yaml’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘mvtnorm’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ape’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘xml2’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘later’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘webutils’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘rappdirs’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘zip’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rhdf5lib’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘stringi’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘purrr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘fs’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘data.table’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘fastmatch’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘protolite’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘plyr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘graph’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘limma’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘tweenr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘tzdb’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘highr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘promises’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘rhdf5filters’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘systemfonts’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘pcaPP’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘robustbase’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘igraph’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rook’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘stringr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘GOSemSim’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘yulab.utils’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘Rgraphviz’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘fgsea’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘vroom’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ggfun’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ggplotify’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ggforce’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘graphlayouts’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘knitr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘reshape2’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘httpuv’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘rhdf5’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘svglite’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ccaPP’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘tidyr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘KEGGgraph’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘tidygraph’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘tidytree’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘readr’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘aplot’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘scatterpie’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘opencpu’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘gson’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘qvalue’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘pathview’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘treeio’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ggraph’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘GEOquery’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘DOSE’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘ggtree’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘phantasus’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘enrichplot’ had non-zero exit status”\n",
      "Warning message in install.packages(...):\n",
      "“installation of package ‘clusterProfiler’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'askpass', 'BiocManager', 'bit', 'bit64', 'bitops', 'cachem',\n",
      "  'cli', 'codetools', 'colorspace', 'crayon', 'curl', 'DBI', 'digest',\n",
      "  'evaluate', 'fansi', 'farver', 'fastmap', 'ggplot2', 'ggrepel', 'glue',\n",
      "  'gtable', 'htmltools', 'IRkernel', 'jsonlite', 'locfit', 'matrixStats',\n",
      "  'munsell', 'nlme', 'openssl', 'pbdZMQ', 'Rcpp', 'RcppArmadillo', 'RCurl',\n",
      "  'repr', 'rlang', 'RSQLite', 'sys', 'utf8', 'uuid', 'withr', 'XML'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "BiocManager::install(c(\"DESeq2\", \"limma\", \"fgsea\", \"AnnotationDbi\", \"org.Mm.eg.db\", \"phantasus\", \"clusterProfiler\", \"pathview\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9f8ac317-6d36-4183-878a-e02996f6a00b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘bit’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘bitops’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘cli’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘colorspace’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘curl’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘digest’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘fansi’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘farver’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘fastmap’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘glue’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘jsonlite’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘locfit’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘matrixStats’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘nlme’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘pbdZMQ’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘Rcpp’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘rlang’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘sys’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘utf8’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘uuid’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘XML’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘askpass’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘bit64’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘cachem’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘htmltools’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘RcppArmadillo’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘RCurl’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘openssl’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘RSQLite’ had non-zero exit status”\n",
      "Warning message in install.packages(update[instlib == l, \"Package\"], l, contriburl = contriburl, :\n",
      "“installation of package ‘ggrepel’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "update.packages(ask = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cde4504f-8029-43a9-b40f-8beee4c504d7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
